src/write_kosak.c

```diff
@@ -0,0 +1,2159 @@
+
+static int wg_curl_get_or_post(
+    char *response_buffer, size_t response_buffer_size,
+    const char *url, const char *header0, const char *header1,
+    const char *body) {
+  const char *get_or_post_tag = body == NULL ? "GET" : "POST";
+  ERROR("Doing: %s request: url %s, header0 %s, header1 %s, body %s",
+        get_or_post_tag, url, header0, header1, body);
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    ERROR("write_kosak: curl_easy_init failed");
+  }
+  const char *collectd_useragent = COLLECTD_USERAGENT;
+  struct curl_slist *headers = NULL;
+  if (header0 != NULL) {
+    headers = curl_slist_append(headers, header0);
+  }
+  if (header1 != NULL) {
+    headers = curl_slist_append(headers, header1);
+  }
+  wg_curl_write_ctx_t write_ctx = {
+     .data = response_buffer,
+     .size = response_buffer_size
+  };
+
+  curl_easy_setopt(curl, CURLOPT_URL, url);
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, collectd_useragent);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+  if (body != NULL) {
+    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body);
+  }
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &wg_curl_write_callback);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &write_ctx);
+
+  int result = curl_easy_perform(curl);
+  if (result != CURLE_OK) {
+    WARNING("write_kosak: curl_easy_perform() failed: %s",
+            curl_easy_strerror(result));
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl);
+    return -1;
+  }
+  write_ctx.data[0] = 0;
+  WARNING("This is the result from curl: %s", response_buffer);
+  if (write_ctx.size < 2) {
+    WARNING("write_kosak: The buffer overflowed.");
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl);
+    return -1;
+  }
+  curl_slist_free_all(headers);
+  curl_easy_cleanup(curl);
+  return 0;
+}
+
+static size_t wg_curl_write_callback(char *ptr, size_t size, size_t nmemb,
+                                     void *userdata) {
+  wg_curl_write_ctx_t *ctx = userdata;
+  if (ctx->size == 0) {
+    return 0;
+  }
+  size_t requested_bytes = size * nmemb;
+  size_t actual_bytes = requested_bytes;
+  if (actual_bytes >= ctx->size) {
+    actual_bytes = ctx->size - 1;
+  }
+  memcpy(ctx->data, ptr, actual_bytes);
+  ctx->data += actual_bytes;
+  ctx->size -= actual_bytes;
+
+  // We lie about the number of bytes successfully transferred so that curl
+  // doesn't freak out and return an error to our caller. Our caller is keeping
+  // track of buffer consumption so it will independently know if the buffer
+  // filled up.
+  return requested_bytes;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// OAuth2 submodule.
```

Comment:
And I think a util_oauth would be fine.  Maybe the creds should just be in the same.  Not sure.